homeassistant/components/home_connect/__init__.py

```diff
@@ -251,6 +263,13 @@ async def async_service_start_program(call):
         schema=SERVICE_PROGRAM_SCHEMA,
     )
 
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_STOP_PROGRAM,
+        async_service_stop_program,
+        schema=SERVICE_COMMAND_SCHEMA,
+    )
```

Comment:
Wait, programs are switches already?

Switches already have services? Why do we need these?

Reply:
@frenck

The stop is an own (DELETE) REST endpoint see https://api-docs.home-connect.com

It just stop the active Programm when the machine is running. It’s not switching a Programm. Because if you run and switch the Programm it will not stop just pause. That’s a different.

If you want to stop you also need no switch or the active Programm this is just handled by the home connect backend.

Pause/Resume works different.

Reply:
@frenck still seems to be right as what you are trying to implement seems to already be part of this [switch](https://github.com/home-assistant/core/blob/39d9a5cda55d4a42c40cdaadaeca496b74dc8ffb/homeassistant/components/home_connect/switch.py#L47)

Reply:
No see here, it's different. Also if you not use the switch.
https://apiclient.home-connect.com/?client_id=YOUR_CLIENT_ID#/programs/stop_program

Reply:
Example:
![Uploading Bildschirmfoto 2023-04-22 um 14.17.35.png…]()

For the button stop, which service can i use?

There is just `home_connect.pause_program`, `home_connect.resume_program`, `home_connect.start_program`
and after this PR also `home_connect.stop_program`

Reply:
So, even if this is the case:

> I just have a button to stop the service, but currently there is no service for that

It should not be a service, but a button.

Either case of the discussion, this PR isn't correct.

Reply:
Why is the resume and pause a service?
For me it make no sense that everything is a service but the stop not, why is the stop missing?

Reply:
> Why is the resume and pause a service?

History maybe? But that doesn't mean we allow/should introduce more things using incorrect constructs.

> The button is just one case, what do you use for a script or automation etc.?

The button.

Reply:
> History maybe? But that doesn't mean we allow/should introduce more things using incorrect constructs.

mhm maybe? But then we can change all together, see no big deal for that.

I have no idea why the pause/resume/start is not a button. I see its currently a service, so for me it was just consistent to do the same.

But also okay todo all with a button, but why do not create a issue for that? Then its consistent...

Reply:
> I have no idea why the pause/resume/start is not a button.

Probably history before we had button entities.

> I see its currently a service, so for me it was just consistent to do the same.

Agree, let's make them all buttons instead.

> but why do not create a issue for that?

It isn't an issue? 🤷

Reply:
> Probably history before we had button entities.

I cannot say it, i use HA since two weeks :).


> It isn't an issue? 🤷
No i meant create a github issue to change all to button entities.
We have two option:
a) Merge it, create a github ticket to change everything to button
b) Decline the PR and make button entities

Or what is your idea?

Reply:
> No i meant create a github issue to change all to button entities.

It is not an issue, so that sounds unneeded.

> Or what is your idea?

Already lined out above

> As a good path forward, I would recommend making a separate PR to migrate/add buttons for the existing services (with deprecation). Once that one is done/merged, we can update this one to match.

Reply:
You edited during my post ;)

Well, as I said, I can just make a new one and add stop there as well.
I would have just merged it briefly and then made all button entities in a new one, but ok.

I can't tell you how much effort it is cause I'm new. Can you give me a code example for button entities?

Reply:
Hey both, I'm getting the impression this exchange is becoming a bit unnecessarily controversial :relaxed: 

Yes, stopping the active program, which would be a service without any data, can be elegantly solved with a button. And yes, I think the reason there are pause/resume services is because that was implemented before there was a button. So @frenck thanks for pointing out this is the preferred way of doing things, but I would not complain to @MelleD for using "incorrect constructs" since I think it is understandable why he originally implemented it like that. Thanks @MelleD for your efforts and @frenck for the review!

Concerning turning the existing services into buttons: just to be sure, we agree that this applies only to `pause_program` and `resume_program`, right? Because all other services have associated data that doesn't make sense to trigger with a simple button.

Reply:
https://developers.home-assistant.io/docs/core/entity/button/

Reply:
> Concerning turning the existing services into buttons: just to be sure, we agree that this applies only to `pause_program` and `resume_program`, right? Because all other services have associated data that doesn't make sense to trigger with a simple button.

Button entities cannot provide additional data in their calls. It is just a single "press" action without parameters. So it indeed aims to replace the services that just target a device or entity without any parameters.

Reply:
Yes you are right, i think start is not possible as button, because you need the program and optional the temperature and maybe other options for starting.

Reply:
@frenck https://github.com/home-assistant/core/pull/91923